File: api/colour-mapping.py
import pandas as pd

from api._lib.blob_io import load_df, save_df
from api._lib.core.colour_mapping import (
    add_suggestions,
    load_colour_mapping,
    mapping_breakdown,
    merge_new_mappings,
    unmapped_colours,
)
from api._lib.core.feed_processor import apply_column_selection, apply_filters
from api._lib.handler_utils import BaseHandler


class handler(BaseHandler):
    def do_GET(self):
        try:
            base_map = load_colour_mapping()
            self._send_json(200, {"rows": [] if base_map is None else base_map.to_dict(orient="records")})
        except Exception as e:
            self._send_json(500, {"error": str(e)})

    def do_POST(self):
        try:
            body = self._json_body()
            raw_df = load_df(body["rawDfUrl"])
            kept = apply_column_selection(raw_df, body.get("keepMap", {}))
            filtered = apply_filters(kept, body.get("filters", {}))
            colour_col = body.get("colourCol")
            base_map = load_colour_mapping()

            if base_map is None or not colour_col or colour_col not in filtered.columns:
                return self._send_json(400, {"error": "Colour mapping unavailable"})

            overrides = body.get("overrides", [])
            new_rows = pd.DataFrame([
                {
                    "product_colour": r["productColour"].strip().lower(),
                    "generic_colour": r["genericColour"].strip().lower(),
                }
                for r in overrides if r.get("productColour") and r.get("genericColour")
            ])
            updated_map = (
                merge_new_mappings(base_map, new_rows.drop_duplicates(subset=["product_colour"]))
                if not new_rows.empty
                else base_map.copy()
            )
            colour_map_url = save_df(updated_map, "colour_map")

            currently_mapped, newly_mapped, unmapped_count, eligible_count, pct_mapped, *_ = mapping_breakdown(
                filtered, colour_col, base_map, updated_map
            )
            allowed_generic = sorted(updated_map["generic_colour"].dropna().unique().tolist())
            unmapped = add_suggestions(unmapped_colours(filtered, colour_col, updated_map), allowed_generic)

            self._send_json(200, {
                "colourMapBlobUrl": colour_map_url,
                "breakdown": {
                    "currentlyMapped": currently_mapped,
                    "newlyMapped": newly_mapped,
                    "unmapped": unmapped_count,
                    "eligible": eligible_count,
                    "pctMapped": pct_mapped,
                },
                "unmapped": [
                    {
                        "productColour": row["Unmapped Colour"],
                        "suggestion": row.get("Suggestion", ""),
                        "productCount": int(row["Product Count"]),
                    }
                    for _, row in unmapped.iterrows()
                ],
                "allowedGeneric": allowed_generic,
            })
        except Exception as e:
            self._send_json(500, {"error": str(e)})

File: api/colour-mapping.ts
import type { VercelRequest, VercelResponse } from '@vercel/node'
import { sendError } from '../api/_lib/handlerUtils'
import { applyColumnSelection, applyFilters } from '../api/_lib/core/feedProcessor'
import {
  loadBaseColourMap, mergeNewMappings, mappingBreakdown,
  unmappedColours, addSuggestions,
} from '../api/_lib/core/colourMapping'
import { loadRows } from '../api/_lib/sessionStore'
import { db, colourMappings } from '../api/_lib/db/index'
import { eq } from 'drizzle-orm'

export default async function handler(req: VercelRequest, res: VercelResponse) {
  try {
    if (req.method === 'GET') {
      const baseMap = loadBaseColourMap()
      const rows = [...baseMap.entries()].map(([productColour, genericColour]) => ({
        productColour, genericColour,
      }))
      return res.status(200).json({ rows })
    }

    if (req.method === 'POST') {
      const { rawDfUrl, session: sessionId, keepMap, filters = {}, colourCol, overrides = [] } = req.body as {
        rawDfUrl: string
        session: string
        keepMap?: Record<string, boolean>
        filters?: Record<string, string[]>
        colourCol: string
        overrides?: { productColour: string; genericColour: string }[]
      }

      let rows = await loadRows(rawDfUrl)
      if (keepMap) rows = applyColumnSelection(rows, keepMap)
      if (Object.keys(filters).length) rows = applyFilters(rows, filters)

      const baseMap = loadBaseColourMap()

      // Persist overrides to Neon (colourMapping core uses snake_case keys internally)
      const overridesSnake = overrides.map(o => ({ product_colour: o.productColour, generic_colour: o.genericColour }))
      if (overridesSnake.length && sessionId) {
        await db.delete(colourMappings).where(eq(colourMappings.sessionId, sessionId))
        await db.insert(colourMappings).values(
          overridesSnake.map(o => ({ sessionId, productColour: o.product_colour, genericColour: o.generic_colour }))
        )
      }

      const updatedMap = mergeNewMappings(baseMap, overridesSnake)
      const bdRaw = mappingBreakdown(rows, colourCol, baseMap, updatedMap)
      const unmappedRaw = unmappedColours(rows, colourCol, updatedMap)
      const allowedGeneric = [...new Set([...updatedMap.values()])].sort()
      const withSuggestions = addSuggestions(unmappedRaw, allowedGeneric)

      return res.status(200).json({
        colourMapBlobUrl: '',
        breakdown: {
          currentlyMapped: bdRaw.currently_mapped,
          newlyMapped: bdRaw.newly_mapped,
          unmapped: bdRaw.unmapped,
          eligible: bdRaw.eligible,
          pctMapped: bdRaw.pct_mapped,
        },
        unmapped: withSuggestions.map(r => ({
          productColour: r.product_colour,
          suggestion: r.suggestion,
          productCount: r.count,
        })),
        allowedGeneric,
      })
    }

    res.status(405).json({ error: 'Method not allowed' })
  } catch (err) {
    sendError(res, err)
  }
}
